core/src/wsv.rs

```diff
@@ -227,18 +244,40 @@ impl<W: WorldTrait> WorldStateView<W> {
         self.metrics.block_height.inc();
     }
 
-    /// Returns blocks after hash
-    pub fn blocks_after(
+    // TODO: There could be just this one method `blocks` instead of `blocks_from_height` and
+    // `blocks_after_height`. Also, this method would return references instead of cloning
+    // blockchain but comes with the risk of deadlock if consumer of the iterator stores
+    // references to blocks
+    ///// Returns iterator over blockchain blocks
+    /////
+    ///// **Locking behaviour**: Holding references to blocks stored in the blockchain can induce
+    ///// deadlock. This limitation is imposed by the fact that blockchain is backed by [`Dashmap`]
+    //pub fn blocks(
+    //    &self,
+    //) -> impl Iterator<Item = impl Deref<Target = VersionedCommittedBlock> + '_> + '_ {
+    //    self.blocks.iter()
+    //}
```

Comment:
I'd like to get some input on this

Reply:
Personally, I'm on the side of caution: we already have too many potential sources of deadlock. Maybe after we figured out what causes Iroha to deadlock under stress (and sometimes in `7_peers_1_fault`), we could experiment with optimisations. Since Iroha is operating in-memory, I think this sort of optimisation would be extremely useful.

Reply:
did your opinion change about this?

Reply:
I could open an issue and put this code example there

Reply:
I agree with the fact that holding an iter with refs is dangerous. We can create a separate test as part of the issue that you mention, and check the behavior of the iterator and dashmap under similar conditions (e.g. trying to put new blocks into dash map while having a ChainIterator being consumed by some block stream).

Reply:
https://github.com/hyperledger/iroha/issues/1737